### ISSUE TITLE
removeRow can cause java.lang.IndexOutOfBoundsException

### ISSUE THREAD

**Issue by nmc9** [evrencoskun/TableView#52]
Using `ITableViewListener` I added a delete button for every record in the table that uses the row index to call `removeRow().`After calling `removeRow()` though the `ITableViewListener` doesn't update the indexes of the table. 

So for example
if you remove row 1 it will remove row 1, but if you then try to remove the new first row (originally row 2) it will delete the row after (originally row 3).
I image the same issue exists for `removeColumn() `but i haven't tested it

**Comment by thearaks** [comment]
Same issue here but with
`mReportListAdapter?.removeRowRange(0, mTableRows.size)`
The table has 2 element and removeRowRange is called with 0, 2 as parameters.

The app crashes with this stack trace:
````
java.lang.IndexOutOfBoundsException: Invalid index 1, size is 1
     at java.util.ArrayList.throwIndexOutOfBoundsException(ArrayList.java:255)
     at java.util.ArrayList.remove(ArrayList.java:403)
     at com.evrencoskun.tableview.adapter.recyclerview.AbstractRecyclerViewAdapter.deleteItemRange(AbstractRecyclerViewAdapter.java:76)
     at com.evrencoskun.tableview.adapter.AbstractTableAdapter.removeRowRange(AbstractTableAdapter.java:171)
````

**Comment by evrencoskun** [comment]
Hi @ncm9,

Sorry for my late response. Contrary to what you may think, TableView actually updates the actual row header index. However, the data of row Header isn't changed what you expect. 

I mean, if you try
   
```java
   @Override
    public void onRowHeaderClicked(@NonNull RecyclerView.ViewHolder rowHeaderView, int **row**) {
           // assume that we have already removed 4. row index.
           // After that, when we try to remove 6. row index.
           //  you can see **row** value ( from this function) won't be returned as 6. 
           // it will return as 5 ( because of 4. is already removed) as we expected. 
           int realIndex = row;
           // Besides, rowHeaderView view holder also gives you the exact index which is the same as row value.
           realIndex = rowHeaderView.getAdapterPosition();
           // However, the data isn't changed. 
           int rowHeaderText = ((RowHeaderViewHolder)rowHeaderView).row_header_textview.getText();
    }

```

So That means, is removeRow method working wrong?  
No, It works perfectly for hiding row process. 


So what the solution? 

when you remove the desired row position of TableView, you can remove the row header which has the largest index.

For example; 

        - 1.row header --------------------- (1.cell items)
        - 2.row header --------------------- (2.cell items)
        - 3.row header --------------------- (3.cell items)

when you remove the 2. row of the TableView, you should remove 2.cell items like removeRow method. However, you also need to remove 3.row header of the RowHeaderRecyclerView. 


```java
removeRowByUpdatingRowHeaderData(int rowPosition){
  // Delete cell items which has located on desired row position
  myTableView.getAdapter().getCellRecyclerViewAdapter().deleteItem(rowPosition);
  
 // Get row header recyclerView adapter
  AbstractRecyclerViewAdapter rowRecyclerViewAdapter = 
 myTableView.getAdapter().getRowHeaderRecyclerViewAdapter();
 // Remove the row header which has the largest index
 rowRecyclerViewAdapter.deleteItem(rowRecyclerViewAdapter.getItemCount() - 1);
}
```

**Comment by nmc9** [comment]
I tried implementing your code to fix the issue and this is what I got.
When I delete row index 4 it does Log row index 4 when the listener is called. Then it deletes row 4. No problem yet. But when I call the method again on row6( the previous row 6 which because of the delete is row 5 now). The listener says the 6th row is being selected when in reality this is now the 5th row.

I don’t know if it makes a difference, but your code is checking the rowHeader clicked whereas mine is checking the cellClicked because the delete button is in one of my cells.


I have been trying to visualize the problem and make sure I explain it right so here is the basic rundown of what is happening:
OK so my table when I start will look something like this
(click to delete row n) is just using the cell click listener to figure out the row to delete when clicked.
```
______| 1.col   | 2.col			 | 
1.row   | 1.cell  | (click to delete row 1)	 |
2.row   | 2.cell  | (click to delete row 2)	 |
3.row   | 3.cell  | (click to delete row 3)* |
4.row   | 4.cell  | (click to delete row 4)	 |
5.row   | 5.cell  | (click to delete row 5)	 |
6.row   | 6.cell  | (click to delete row 6)	 |
```

When I click for example delete row 3* the table changes to
```
______| 1.col   | 2.col			 | 
1.row   | 1.cell  | (click to delete row 1)	 |
2.row   | 2.cell  | (click to delete row 2)	 |
3.row   | 4.cell  | (click to delete row 4)**   |
4.row   | 5.cell  | (click to delete row 5)	 |
5.row   | 6.cell  | (click to delete row 6)	 |
```
It deletes the right row but the references to the rows to delete is not updated.
Notice how the listener doesn’t go down because ideally the table would look like this
```
______| 1.col   | 2.col			 | 
1.row   | 1.cell  | (click to delete row 1)	 |
2.row   | 2.cell  | (click to delete row 2)	 |
3.row   | 4.cell  | (click to delete row 3)   |
4.row   | 5.cell  | (click to delete row 4)	 |
5.row   | 6.cell  | (click to delete row 5)	 |
```

So when I click on the delete button in row 3** it actually deletes the next row (row 4)
```
______| 1.col   | 2.col			 | 
1.row   | 1.cell  | (click to delete row 1)	 |
2.row   | 2.cell  | (click to delete row 2)	 |
3.row   | 4.cell  | (click to delete row 4)   |
4.row   | 5.cell  | (click to delete row 5)*** |
5.row   | 6.cell  | (click to delete row 6)    |
```
This leaves me with a table where the listener is completely incarnate is it’s after the delete.
So what it seems like I need is for the listener to update when the remove is called.

**Comment by nmc9** [comment]
Ok thank you I believe I found what was causing my problems. I used your code `myTableView.getAdapter().getCellRecyclerViewAdapter().notifyDataSetChanged().` to update the adapter and after that everything seems to be working correctly.
I had tried calling `myTableView.notifyDataSetChanged() `but it didn't seem to update the adapter in the right way. I looked around and its seems to be because calling notifyDataSetChanged() directly on the tableAdapter calls notify**Cell**DataSetChanged(); instead of notifyDataSetChanged().
```
public final void notifyDataSetChanged() {
        mColumnHeaderRecyclerViewAdapter.notifyDataSetChanged();
        mRowHeaderRecyclerViewAdapter.notifyDataSetChanged();
        mCellRecyclerViewAdapter.notifyCellDataSetChanged();
    }
```
So If i called notifyDataSetChanged on the Cell adapter all is well.

Thank you so much for your time

PS: That is a really nice graphic of the table visually broken down by recyclerviews. It really helped when I was looking around and debugging. You might want to think about adding that somewhere in the documentation.

**Comment by evrencoskun** [comment]
Hi, @thearaks 

The bug has been fixed in 0.8.7 version. Please check the case using the version.

Hi @nmc9, 

TableViewAdapter has a new method for the request like yours.

```java
void removeRow(int rowPosition, boolean updateRowHeader) 
void removeRowRange(int rowPositionStart, int itemCount, boolean updateRowHeader)
```

So you can just use the above methods to remove the desired row.

**Comment by nmc9** [comment]
Thank you very much. Just tested real quick and it's working perfectly
